sdk/containerregistry/Azure.Containers.ContainerRegistry/src/Blob/ContainerRegistryBlobClient.cs

```diff
@@ -282,6 +240,51 @@ public virtual async Task<Response<UploadManifestResult>> UploadManifestAsync(St
             }
         }
 
+        private async Task<Response<UploadManifestResult>> UploadManifestInternalAsync(Stream stream, UploadManifestOptions options, bool async, CancellationToken cancellationToken)
+        {
+            Debug.Assert(stream is MemoryStream, "Should only be called on internally allocated, seekable streams.");
+
+            string digest = OciBlobDescriptor.ComputeDigest(stream);
+            string tagOrDigest = options.Tag ?? digest;
+
+            ResponseWithHeaders<ContainerRegistryCreateManifestHeaders> response = async ?
+                // TODO: media type should be configurable to support non-OCI types.
+                await _restClient.CreateManifestAsync(_repositoryName, tagOrDigest, stream, ManifestMediaType.OciManifest.ToString(), cancellationToken).ConfigureAwait(false) :
+                _restClient.CreateManifest(_repositoryName, tagOrDigest, stream, ManifestMediaType.OciManifest.ToString(), cancellationToken);
+
+            if (!ValidateDigest(digest, response.Headers.DockerContentDigest))
+            {
+                throw _clientDiagnostics.CreateRequestFailedException(response,
+                    new ResponseError(null, "The digest in the response does not match the digest of the uploaded manifest."));
```

Comment:
Isn't ResponseError meant for service returned errors? Why not just pass the message?